### DISCUSSION TITLE
Solutions for the train-kissing problem

### DISCUSSION THREAD

**Discussion by ghost** [voxeltycoon/issues#527] [General]
Not knowing the algorithms behind VT, I've searched for more than one way to resolve the problem.

SOLUTION 1: "the long one"
We're going to analyze the standard station prone to a train-kissing problem (3 platforms, 2 one-way tracks & 1 two-way track)
![InkedScreenshot (8)_LI](https://user-images.githubusercontent.com/52663433/112319830-bc859000-8cae-11eb-919c-07a683c230b7.jpg)

The phenomenon takes place when two trains approach stations which have two-way tracks at the same time.
Trains cannot see behind the signal in front of them, but maybe signals could be improved to analyze the situation in the following sections.
First of all, we must know the basic signaling system in a station: 1 pre-signal to protect inbound junctions; 1 signal indicating the beginning of the platform; 1 pre-signal to indicate the end of the platform and to protect the outgoing junctions; 1 signal indicating the end of the railyard (they are shown in the previous pic).

Now, the logic tree where we're going to take in exam the train coming from the right and its route:
(1) THE SIGNAL IN FRONT OF THE TRAIN IS A PRE-SIGNAL
     No = ;
     Yes = next point;
(2) IT'S PROTECTING AT LEAST 1 TWO-WAY PLATFORM
     No = ;
     Yes = next point;
(3) IT'S PROTECTING TWO-WAY & ONE-WAY PLATFORMS
     No = next point (the station has only two-way platforms);
     Yes = [send the train to the one-way platform];

The problem would seem resolved in this base case but we must consider other possibilities.
(4) THE THIRD SIGNAL (end of the platform) IS RED
     No = ;
     Yes = next point;

From here we're going to consider only sections between the third signal and the following pre-signals facing in the opposite direction (note: the 3° reversed pre-signal crosses the green section but there are not tracks joined with the station->not relevant pre-signal)
![InkedScreenshot (11)_LI](https://user-images.githubusercontent.com/52663433/112324148-c0b3ac80-8cb2-11eb-94ca-26242147e2da.jpg)
(5) IS THE OCCUPYING TRAIN ENTERING OR LEAVING THE STATION?
     Entering = next point;
     Leaving = ;

Dividing junctions in two types (according to their direction), we will have...
![Screenshot (13)_LI](https://user-images.githubusercontent.com/52663433/112328082-46852700-8cb6-11eb-8501-d2d9d4684238.jpg)
1° case: both tracks lead to the same platform (that must be considered occupied).
2° case: the track leads to two different platforms (we can't know where the train is going).

It's necessary to reduce the green section to an even smaller block, the one between the 3rd signal and the nose of the train.
(6) THIS SECTION INCLUDES JUNCTIONS WITH THE DOVETAIL TOWARDS THE STATION
     No = [the platform to which it leads must be considered occupied];
     Yes = next point (we can't know where the train is going until it passes the last switch with its dovetail facing the station);

In this picture the train is approaching the first junction that leads to every possible platform, so they are all set as "waiting for info".
![InkedScreenshot (2)_LI](https://user-images.githubusercontent.com/52663433/112309330-81ca2a80-8ca3-11eb-9c84-20f91a9a306e.jpg)

After the first junction, some platforms can be considered available to receive trains from the opposite direction. But there are other exchanges...
![InkedScreenshot (3)_LI](https://user-images.githubusercontent.com/52663433/112309366-8c84bf80-8ca3-11eb-9829-7829f450962c.jpg)

Now that we know the final destination, we can set it as "occupied" even before the train enters it and set the others as "clear"
![InkedScreenshot (4)_LI](https://user-images.githubusercontent.com/52663433/112309377-90184680-8ca3-11eb-9308-3fc43b3f708b.jpg)

PRO: no more train-kissing 
CONS: the second train is forced to wait for "clear" tracks at the 1° signal (outside the railyard), if the logic tree reaches the last point / this process has to be replicated every time trains meet a pre-signal and it could require a large amount of memory (Idk).

SOLUTION 2: "the light one"
Equip each platform with a variable that communicates when the platform is booked by one of the trains that have just entered the railyard. So, trains consider occupied a platform although its own signal is green.

PRO: also here, no more train-kissing
CONS: none (as far as I know)

**Comment by IDragonfire** [comment]
Another idea: Link blocks together.
If a link part is busy the other part is also blocked. 
Does this not solve some issues?
